### PR TITLE
Support global variables, alloca/glbvar's address should be non-zero

### PR DESCRIPTION
**Pull request by aqjune**
This patch adds support for global variables, and makes alloca/glbvar's address never be zero.
malloc's address is not addressed here, as malloc() can return null in runtime. (#110 )

Global variable support is done as follows:
- A global variable in src and tgt are given the same block id , and this information is passed as
```
    sym_exec(src_state);
    tgt_state.setGlobalVarBidsFrom(src_state);
    sym_exec(tgt_state);
```
at TransformVerify::verify().
- To give the block id precisely, `Memory::alloc` now can take an optional argument that specifies which bid to use. It is caller's responsibility to give a unique bid.

LLVM unit tests result:
- Failures: 133 -> 173 (+40), around +30 tests are global (test/Transforms/SafeStack , test/Transforms/GlobalOpt). For other tests, investigation is needed.
- Running time: 1197 -> 296 secs; speedup exists, which is a bit weird :/

### CODE REVIEW

**Comment by aqjune** [comment]
Writing to const global variable should raise UB, but it is not addressed in this PR.

File: llvm_util/utils.h [hunk 1]
```diff
@@ -5,6 +5,7 @@
 
 #include <ostream>
 #include <string>
+#include "llvm/IR/DataLayout.h"
```
Comment (nunoplopes) [inline llvm_util/utils.h]: don't include this header here; add a forward decl below

File: llvm_util/utils.h [hunk 2]
```diff
@@ -29,7 +30,7 @@ IR::Type& get_int_type(unsigned bits);
 IR::Type* llvm_type2alive(const llvm::Type *ty);
 
 IR::Value* make_intconst(uint64_t val, int bits);
-IR::Value* get_operand(llvm::Value *v);
+IR::Value* get_operand(llvm::Value *v, const llvm::DataLayout &DL);
```
Comment (nunoplopes) [inline llvm_util/utils.h]: this makes all calls more verbose. it's better to pass this to init_llvm_utils() below.

File: llvm_util/utils.cpp [hunk 3]
```diff
@@ -51,7 +53,7 @@ string value_name(const llvm::Value &v) {
   if (!v.getName().empty())
     return name = '%' + v.getName().str();
   return name = v.getType()->isVoidTy() ? "<void>"
-                                        : '%' + to_string(value_id_counter++);
+                                        : "%" + to_string(value_id_counter++);
```
Comment (nunoplopes) [inline llvm_util/utils.cpp]: why this change?

File: ir/value.cpp [hunk 4]
```diff
@@ -5,7 +5,6 @@
 #include "smt/expr.h"
 #include "util/compiler.h"
 #include "util/config.h"
-
```
Comment (nunoplopes) [inline ir/value.cpp]: ...

File: ir/value.cpp [hunk 5]
```diff
@@ -88,7 +87,32 @@ StateValue NullPointerValue::toSMT(State &s) const {
...
+  unsigned glbvar_bid;
+  if (s.hasGlobalVarBid(getName(), glbvar_bid))
+    // Use the same block id that is used by src
+    ptrval = s.getMemory().alloc(sizeexpr, align, Memory::GLOBAL, glbvar_bid);
+  else {
+    ptrval = s.getMemory().alloc(sizeexpr, align, Memory::GLOBAL, nullopt,
+                                 &glbvar_bid);
+    s.addGlobalVarBid(getName(), glbvar_bid);
+  }
+  if (isconst && initval) {
```
Comment (nunoplopes) [inline ir/value.cpp]: is there any reason to store the init val if !isconst?  if not, we shouldn't store it at all.

File: ir/value.cpp [hunk 6]
```diff
@@ -88,7 +87,32 @@ StateValue NullPointerValue::toSMT(State &s) const {
...
+void GlobalVariable::print(ostream &os) const {
+  UNREACHABLE();
+}
+
+StateValue GlobalVariable::toSMT(State &s) const {
+  auto sizeexpr = expr::mkInt(allocsize, 64);
+  expr ptrval;
+  unsigned glbvar_bid;
+  if (s.hasGlobalVarBid(getName(), glbvar_bid))
+    // Use the same block id that is used by src
```
Comment (nunoplopes) [inline ir/value.cpp]: this should assert(!is_src)

File: tools/transform.cpp [hunk 7]
```diff
@@ -310,6 +310,7 @@ Errors TransformVerify::verify() const {
 
   try {
     sym_exec(src_state);
+    tgt_state.setGlobalVarBidsFrom(src_state);
```
Comment (nunoplopes) [inline tools/transform.cpp]: maybe call it copyGlobalVarsFromSrc() to make it explicit

File: tests/alive-tv/glb-reorder.src.ll [hunk 8]
```diff
@@ -0,0 +1,9 @@
+@A = global i32 0
+@B = global i32 0
+
+define i32 @foo() {
+  %l = load i32, i32* @A
+  store i32 0, i32* @B
```
[the remark this answers is not in the document]
Comment (aqjune) [reply tests/alive-tv/glb-reorder.src.ll]: Updated. :)

File: llvm_util/utils.h (hunk 2 above)
Comment (aqjune) [reply llvm_util/utils.h]: Yep, updated.

File: llvm_util/utils.cpp (hunk 3 above)
Comment (aqjune) [reply llvm_util/utils.cpp]: It was because visual studio code couldn't syntax highlight the following code if it was '%', just a silly reason :/ Undoed the change.

File: ir/value.cpp (hunk 4 above)
Comment (aqjune) [reply ir/value.cpp]: Sorry :P updated

File: ir/value.cpp (hunk 5 above)
Comment (aqjune) [reply ir/value.cpp]: Nope, if a global variable is not constant, it is not necessary to store its value because a caller may change that. This if block states that it is already ignoring the initial value.

File: ir/value.cpp (hunk 6 above)
Comment (aqjune) [reply ir/value.cpp]: Added

File: tools/transform.cpp (hunk 7 above)
Comment (aqjune) [reply tools/transform.cpp]: Updated

File: ir/state.cpp [hunk 9]
```diff
@@ -13,9 +13,6 @@ namespace IR {
 
 State::State(const Function &f, bool source)
   : f(f), source(source), memory(*this) {
-  predecessor_data[&f.getFirstBB()].try_emplace(nullptr,
-                                                DomainTy(true, set<expr>()),
-                                                *this);
```
Comment (aqjune) [inline ir/state.cpp]: The reason why I removed the predecesor_data for FirstBB was that it was storing an empty memory which made State::startBB(FirstBB) ovewrite the initial memory which contained initial values of const global vars. I made predecessor_data[FirstBB] store nothing and updated State::startBB.

**Comment by aqjune** [comment]
Updated PR.
llvm unit tests: 108 fails (master) -> 133 fails. It seems all fails are related with global optimizations / X86-specific optimizations (safestack thingy) / atomic operations.

File: ir/memory.cpp [hunk 10]
```diff
@@ -248,16 +248,37 @@ pair<expr, vector<expr>> Memory::mkInput(const char *name) {
...
+  // It is allocated as a local block if blockKind is heap or stack.
+  bool is_local = blockKind != GLOBAL;
+
+  Pointer p(*this, bid, is_local);
   state->addPre(p.is_aligned(align));
 
   expr size_zextOrTrunced = size.zextOrTrunc(bits_size_t);
   state->addPre(p.block_size() == size_zextOrTrunced);
+  // TODO: If its address space is not 0, its address can be 0.
+  state->addPre(p.get_address() != 0);
```
Comment (nunoplopes) [inline ir/memory.cpp]: wait, I thought the conclusion is that malloc can return null?

File: ir/value.cpp (hunk 5 above)
Comment (nunoplopes) [reply ir/value.cpp]: Then let's remove the `&& initval` bit.

File: llvm_util/utils.cpp [hunk 11]
```diff
@@ -165,6 +170,34 @@ Value* get_operand(llvm::Value *v) {
...
+      // TODO: Array type of a global variable is not supported.
+      return nullptr;
+    } else if (auto st = dyn_cast<llvm::StructType>(gv->getValueType())) {
+      if (st->isOpaque())
+        // TODO: Global variable of opaque type is not supported.
+        return nullptr;
+    }
+    Value *initval = nullptr;
+    if (gv->hasInitializer()) {
+      initval = get_operand(gv->getInitializer());
```
Comment (nunoplopes) [inline llvm_util/utils.cpp]: pls don't the initializer if it isn't const

File: ir/memory.cpp [hunk 12]
```diff
@@ -248,16 +248,37 @@ pair<expr, vector<expr>> Memory::mkInput(const char *name) {
   return { Pointer(*this, offset, local_bid, bid).release(), { var } };
 }
 
-expr Memory::alloc(const expr &size, unsigned align, bool heap) {
-  Pointer p(*this, ++last_bid, true);
+expr Memory::alloc(const expr &size, unsigned align, BlockKind blockKind,
+                   optional<unsigned> bidopt, unsigned *bid_out) {
+  unsigned bid = bidopt ? *bidopt : last_bid;
+  if (last_bid <= bid) {
```
Comment (nunoplopes) [inline ir/memory.cpp]: you do a bump in tgt, and in src bidopt should be null. I would ensure this invariant with:
```
if (!bidopt)
  ++last_bid;
assert(bid < last_bid);
```

File: ir/state.cpp [hunk 13]
```diff
@@ -177,4 +178,33 @@ void State::resetUndefVars() {
   undef_vars.clear();
 }
 
+void State::addGlobalVarBid(const std::string &glbvar, unsigned bid) {
+  assert(glbvar_bids.find(glbvar) == glbvar_bids.end());
+  glbvar_bids[glbvar] = bid;
+}
+
+bool State::hasGlobalVarBid(const std::string &glbvar,
```
Comment (nunoplopes) [inline ir/state.cpp]: no std::

File: ir/state.cpp [hunk 14]
```diff
@@ -177,4 +178,33 @@ void State::resetUndefVars() {
   undef_vars.clear();
 }
 
+void State::addGlobalVarBid(const std::string &glbvar, unsigned bid) {
```
Comment (nunoplopes) [inline ir/state.cpp]: no std::

File: ir/state.cpp [hunk 15]
```diff
@@ -177,4 +178,33 @@ void State::resetUndefVars() {
...
+  assert(glbvar_bids.find(glbvar) == glbvar_bids.end());
+  glbvar_bids[glbvar] = bid;
```
Comment (nunoplopes) [inline ir/state.cpp]: .emplace() is more efficient

File: ir/state.cpp [hunk 16]
```diff
@@ -177,4 +178,33 @@ void State::resetUndefVars() {
   undef_vars.clear();
 }
 
+void State::addGlobalVarBid(const std::string &glbvar, unsigned bid) {
+  assert(glbvar_bids.find(glbvar) == glbvar_bids.end());
```
Comment (nunoplopes) [inline ir/state.cpp]: .count(glbvar) == 0

File: ir/state.cpp [hunk 17]
```diff
@@ -82,26 +79,30 @@ bool State::startBB(const BasicBlock &bb) {
   ENSURE(seen_bbs.emplace(&bb).second);
   current_bb = &bb;
 
+  bool isFirstBB = f.getFirstBB().getName() == bb.getName();
```
Comment (nunoplopes) [inline ir/state.cpp]: instead of this huge change, can we do instead (since none of the code later is relevant):
```
if (f.getFirstBB().getName() == bb.getName())
  return true;
```

File: llvm_util/utils.cpp [hunk 18]
```diff
@@ -165,6 +170,34 @@ Value* get_operand(llvm::Value *v) {
     return ret;
   }
 
+  if (auto gv = dyn_cast<llvm::GlobalVariable>(v)) {
+    auto itr = globalvars.find(gv);
+    if (itr != globalvars.end())
+      return itr->second;
+
+    if (gv->getValueType()->isArrayTy()) {
+      // TODO: Array type of a global variable is not supported.
```
Comment (nunoplopes) [inline llvm_util/utils.cpp]: I don't see any reason for limiting the type. In the worst case, get_operand returns a null expr.

File: ir/memory.cpp (hunk 10 above)
Comment (aqjune) [reply ir/memory.cpp]: I believe this can be addressed by making Malloc::toSMT return `mkIf(random_bool, Memory::alloc(), nullptr)`. I will make a separate PR for this so discussion for its implementation can be done there.

File: ir/state.cpp (hunk 15 above)
Comment (aqjune) [reply ir/state.cpp]: Updated

File: llvm_util/utils.cpp (hunk 18 above)
Comment (aqjune) [reply llvm_util/utils.cpp]: A global array's type is a pointer to array, so llvm_type2alive(v->getType()) at the beginning of get_operand successfully returns IR type, that's why isArrayTy() here is needed.

File: ir/memory.cpp (hunk 12 above)
Comment (aqjune) [reply ir/memory.cpp]: I see, updated.

File: ir/value.cpp (hunk 5 above)
Comment (aqjune) [reply ir/value.cpp]: Const global may not have its initial value if it is extern variable, so initial is still needed. Instead, I'll leave initval only.

File: llvm_util/utils.cpp (hunk 11 above)
Comment (aqjune) [reply llvm_util/utils.cpp]: updated

**Comment by aqjune** [comment]
Addressed comments.
Unit tests: 116 fails -> 141 fails (+25, as described in the previous iteration).
